src/runtime-rs/crates/runtimes/virt_container/src/factory/vm/mod.rs

```diff
@@ -0,0 +1,303 @@
+        }
+
+        // memory_size
+        if conf.memory_info.default_memory == 0 {
+            conf.memory_info.default_memory = default::DEFAULT_QEMU_MEMORY_SIZE_MB;
+        }
+
+        // default_bridges
+        if conf.device_info.default_bridges == 0 {
+            conf.device_info.default_bridges = default::DEFAULT_QEMU_PCI_BRIDGES;
+        }
+
+        // block_device_driver
+        if conf.blockdev_info.block_device_driver.is_empty() {
+            conf.blockdev_info.block_device_driver = default::DEFAULT_BLOCK_DEVICE_TYPE.to_string();
+        } else if conf.blockdev_info.block_device_driver == "virtio-blk"
+            && conf.machine_info.machine_type == "s390-ccw-virtio"
+        {
+            conf.blockdev_info.block_device_driver = "virtio-blk-ccw".to_string();
+        }
+
+        // default_maxvcpus
+        if conf.cpu_info.default_maxvcpus == 0
+            || conf.cpu_info.default_maxvcpus > default::MAX_QEMU_VCPUS
+        {
+            conf.cpu_info.default_maxvcpus = default::MAX_QEMU_VCPUS;
+        }
+
+        Ok(())
+    }
+}
+
+impl VM {
+    // Initializes the QEMU hypervisor for Kata
+    async fn new_hypervisor(config: &VMConfig) -> Result<Arc<dyn Hypervisor>> {
+        let hypervisor: Arc<dyn Hypervisor> = match config.hypervisor_name.as_str() {
+            HYPERVISOR_QEMU => {
+                let h = Qemu::new();
+                h.set_hypervisor_config(config.hypervisor_config.clone())
+                    .await;
+                Arc::new(h)
+            }
+            _ => return Err(anyhow!("Unsupported hypervisor {}", config.hypervisor_name)),
+        };
+        Ok(hypervisor)
+    }
+
+    // Initializes the Kata agent, handling necessary configurations and setup
+    fn new_agent(config: &VMConfig) -> Result<Arc<KataAgent>> {
+        let agent_name = &config.agent_name;
+        let agent_config = config.agent_config.clone();
+
+        match agent_name.as_str() {
+            AGENT_KATA => {
+                let agent = KataAgent::new(agent_config.clone());
+                Ok(Arc::new(agent))
+            }
+            _ => Err(anyhow!("Unsupported agent {}", &agent_name)),
+        }
+    }
+
+    // Create an empty `sandbox_config` structure
+    fn new_empty_sandbox_config() -> SandboxConfig {
+        SandboxConfig {
+            sandbox_id: String::new(),
+            hostname: String::new(),
+            dns: Vec::new(),
+            network_env: SandboxNetworkEnv {
+                netns: None,
+                network_created: false,
+            },
+            annotations: HashMap::default(),
+            hooks: None,
+            state: runtime_spec::State {
+                version: Default::default(),
+                id: String::new(),
+                status: runtime_spec::ContainerState::Creating,
+                pid: 0,
+                bundle: String::new(),
+                annotations: Default::default(),
+            },
+            shm_size: 0,
+        }
+    }
```

Comment:
I prefer to derive `Default` for `SandboxNetworkEnv` and `SandboxConfig`, then use `SandboxConfig::default()`. For example:

```rust
#[derive(Clone, Default)]
pub struct SandboxNetworkEnv {
    pub netns: Option<String>,
    pub network_created: bool,
}
```